evm/src/cpu/kernel/interpreter.rs

```diff
@@ -1083,51 +1167,77 @@ impl<'a> Interpreter<'a> {
         self.push(syscall_info)
     }
 
-    fn set_jumpdest_bit(&mut self, x: U256) -> U256 {
+    fn get_jumpdest_bit(&self, offset: usize) -> U256 {
         if self.generation_state.memory.contexts[self.context()].segments
             [Segment::JumpdestBits.unscale()]
         .content
         .len()
-            > x.low_u32() as usize
+            > offset
         {
             self.generation_state.memory.get(MemoryAddress {
                 context: self.context(),
                 segment: Segment::JumpdestBits.unscale(),
-                virt: x.low_u32() as usize,
+                virt: offset,
             })
         } else {
             0.into()
         }
     }
-    fn run_jump(&mut self) -> anyhow::Result<(), ProgramError> {
-        let x = self.pop()?;
 
-        let jumpdest_bit = self.set_jumpdest_bit(x);
+    pub(crate) fn get_jumpdest_bits(&self, context: usize) -> Vec<bool> {
+        self.generation_state.memory.contexts[context].segments[Segment::JumpdestBits.unscale()]
+            .content
+            .iter()
+            .map(|x| x.bit(0))
+            .collect()
+    }
+
+    fn add_jumpdest_offset(&mut self, offset: usize) {
+        if let Some(jumpdest_table) = self
+            .jumpdest_table
+            .get_mut(&self.generation_state.registers.context)
+        {
+            jumpdest_table.insert(offset);
+        } else {
+            self.jumpdest_table.insert(
+                self.generation_state.registers.context,
+                BTreeSet::from([offset]),
+            );
+        }
+    }
+
+    fn run_jump(&mut self) -> anyhow::Result<(), ProgramError> {
+        let offset = self.pop()?;
 
         // Check that the destination is valid.
-        let x: u32 = x
+        let offset: usize = offset
             .try_into()
             .map_err(|_| ProgramError::InvalidJumpDestination)?;
```

Comment:
There's also a `u256_to_usize` method that returns a `ProgramError`, but `try_into` is fine too.